src/emqtt_ws.erl

```diff
@@ -45,36 +45,48 @@ connect(Host0, Port, Opts, Timeout) ->
                                         retry => 0}),
     case gun:open(Host1, Port, ConnOpts) of
         {ok, ConnPid} ->
+            MRef = monitor(process, ConnPid),
             case gun:await_up(ConnPid, Timeout) of
                 {ok, _} ->
-                    case upgrade(ConnPid, Opts, Timeout) of
+                    case upgrade(ConnPid, MRef, Opts, Timeout) of
                         {ok, StreamRef} -> {ok, {ConnPid, StreamRef}};
                         Error -> Error
                     end;
-                Error -> Error
+                Error ->
+                    demonitor(MRef, [flush]),
+                    Error
             end;
-        Error -> Error
+        Error ->
+            Error
     end.
 
 opts([], Acc) -> Acc;
 opts([{Name, Value} | More], Acc) ->
     opts(More, Acc#{Name => Value}).
 
--spec(upgrade(pid(), list(), timeout())
+-spec(upgrade(pid(), reference(), list(), timeout())
       -> {ok, Headers :: list()} | {error, Reason :: term()}).
-upgrade(ConnPid, Opts, Timeout) ->
+upgrade(ConnPid, MRef, Opts, Timeout) ->
     %% 2. websocket upgrade
     Path = proplists:get_value(ws_path, Opts, "/mqtt"),
     CustomHeaders = proplists:get_value(ws_headers, Opts, []),
     StreamRef = gun:ws_upgrade(ConnPid, Path, ?WS_HEADERS ++ CustomHeaders, ?WS_OPTS),
     receive
         {gun_upgrade, ConnPid, StreamRef, [<<"websocket">>], _Headers} ->
+            demonitor(MRef, [flush]),
             {ok, StreamRef};
         {gun_response, ConnPid, _, _, Status, Headers} ->
+            demonitor(MRef, [flush]),
             {error, {ws_upgrade_failed, Status, Headers}};
         {gun_error, ConnPid, StreamRef, Reason} ->
+            demonitor(MRef, [flush]),
+            {error, {ws_upgrade_failed, Reason}};
+        {'DOWN', MRef, process, ConnPid, {shutdown, Reason}} ->
+            {error, {ws_upgrade_failed, Reason}};
+        {'DOWN', MRef, process, ConnPid, Reason} ->
             {error, {ws_upgrade_failed, Reason}}
```

Comment:
I think we're supposed to receive and handle `{gun_down, ...}` as well, weird that there's no clause for it.

Reply:
Good catch!  That indeed can contain the TLS alert we're looking for.